### PR TITLE
fix: [NGRM] - Strange artefacts in 2D maps #2134

### PR DESCRIPTION
**Pull request by nilscb**

### CODE REVIEW

File: typescript/packages/subsurface-viewer/src/layers/map/fragment.fs.glsl.ts [hunk 1]
```diff
@@ -54,6 +54,14 @@ void main(void) {
    vec4 color = vec4(1.0, 1.0, 1.0,  1.0);;
    float propertyValue = property;
 
+   // This may happen due to GPU interpolation precision causing color artifacts.
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/fragment.fs.glsl.ts:57]: use clam(propertyValue, valueRangeMin, valueRangeMax);

Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/map/fragment.fs.glsl.ts:57]: Fixed.

**Comment by nilscb** [comment]
Added same fix for grid3D layer

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts [hunk 2]
```diff
@@ -26,8 +28,12 @@ uniform bool isColorMapClampColorTransparent;
 vec4 getContinuousPropertyColor (float propertyValue) {
 
    vec4 color = vec4(1.0, 1.0, 1.0, 1.0);
+
+   // This may happen due to GPU interpolation precision causing color artifacts.
+   propertyValue = clamp(propertyValue, valueRangeMin, valueRangeMax);
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts:33]: I will add Leonid also to this code review as he has done lots of chenges to the grid3d fragment shader.

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts [hunk 3]
```diff
@@ -52,9 +58,8 @@ vec4 getContinuousPropertyColor (float propertyValue) {
 vec4 getDiscretePropertyColor (float propertyValue) {
 
    vec4 color = vec4(1.0, 1.0, 1.0, 1.0);
-   float tolerance = (1.0 / colorMapSize) * 0.5;
-     
-   if (propertyValue < colorMapRangeMin - tolerance || propertyValue > colorMapRangeMax + tolerance) {
+    
+   if (propertyValue < colorMapRangeMin || propertyValue > colorMapRangeMax) {
```
Comment (LeonidPolukhin) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts:62]: Have you checked that discrete property grids with clamping are colored correctly and there is no "color fighting"?

Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts]: I added the tolerances back in for this layer.  Im not quite sure why they are needed in the continuous case.

**Review by LeonidPolukhin — APPROVED** [review]

**Review by w1nklr — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (w1nklr) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts]: good

**Review by LeonidPolukhin — APPROVED** [review]

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts [hunk 4]
```diff
@@ -5,29 +5,33 @@ precision highp float;
 
 in vec3 cameraPosition;
 in vec4 position_commonspace;
-in float property;
+flat in float property;
+in float property_interpolated;
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts:9]: Be aware that this means duplication of data sent to the fragment shader, with possible memory impact.

Let's finalize this PR as is, but we might want to create 2 files, for 2 different shaders: one flat version, one regular version...
You can resolve this comment

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts [hunk 5]
```diff
@@ -5,29 +5,33 @@ precision highp float;
...
 uniform sampler2D colormap;
 
+uniform float valueRangeMin;
+uniform float valueRangeMax;
 uniform float colorMapRangeMin;
 uniform float colorMapRangeMax;
 
 uniform bool  isColoringDiscrete;
 uniform float colorMapSize;
+uniform highp int coloringMode;
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts:23]: Does `highp` this really have any impact ? Just to store 0, 1, 2 and 4 ?

**Review by w1nklr — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts [hunk 6]
```diff
@@ -26,6 +28,7 @@ uniform bool isColorMapClampColorTransparent;
 vec4 getContinuousPropertyColor (float propertyValue) {
 
    vec4 color = vec4(1.0, 1.0, 1.0, 1.0);
+
    float x = (propertyValue - colorMapRangeMin) / (colorMapRangeMax - colorMapRangeMin);
    if (x < 0.0 - 1e-4 || x > 1.0 + 1e-4) {
```
[the remark this answers is not in the document]
Comment (w1nklr) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts]: OK, let's keep that for now...

File: typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts (hunk 5 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts:23]: Yes it has an impact. Without it the program does not work . I had to google it. The fragment shader needs this, The vertex shader do not. Even if it is the same uniform.

Comment (w1nklr) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/fragment.fs.glsl.ts:23]: ok

**Review by w1nklr — APPROVED** [review]

**Pull request merged**